pandas/core/arrays/period.py

```diff
@@ -228,6 +303,183 @@ def start_time(self):
+                raise IncompatibleFrequency(msg)
+
+            value = value.ordinal
+        elif isna(value):
+            value = iNaT
+        else:
+            msg = ("'value' should be a 'Period', 'NaT', or array of those. "
+                   "Got '{}' instead.".format(type(value).__name__))
+            raise TypeError(msg)
+        self._data[key] = value
+
+    def take(self, indices, allow_fill=False, fill_value=None):
+        if allow_fill:
+            if isna(fill_value):
+                fill_value = iNaT
+            elif isinstance(fill_value, Period):
+                if self.freq != fill_value.freq:
+                    msg = DIFFERENT_FREQ_INDEX.format(
+                        self.freq.freqstr,
+                        fill_value.freqstr
+                    )
+                    raise IncompatibleFrequency(msg)
+
+                fill_value = fill_value.ordinal
+            else:
+                msg = "'fill_value' should be a Period. Got '{}'."
+                raise ValueError(msg.format(fill_value))
+
+        new_values = algos.take(self._data,
+                                indices,
+                                allow_fill=allow_fill,
+                                fill_value=fill_value)
+
+        return type(self)(new_values, self.freq)
+
+    def isna(self):
+        return self._data == iNaT
+
+    def fillna(self, value=None, method=None, limit=None):
+        # TODO(#20300)
+        # To avoid converting to object, we re-implement here with the changes
+        # 1. Passing `_ndarray_values` to func instead of self.astype(object)
+        # 2. Re-boxing output of 1.
+        # #20300 should let us do this kind of logic on ExtensionArray.fillna
+        # and we can use it.
+
+        if isinstance(value, ABCSeries):
+            value = value._values
+
+        value, method = validate_fillna_kwargs(value, method)
+
+        mask = self.isna()
+
+        if is_array_like(value):
+            if len(value) != len(self):
+                raise ValueError("Length of 'value' does not match. Got ({}) "
+                                 " expected {}".format(len(value), len(self)))
+            value = value[mask]
+
+        if mask.any():
+            if method is not None:
+                func = pad_1d if method == 'pad' else backfill_1d
+                new_values = func(self._ndarray_values, limit=limit,
+                                  mask=mask)
+                new_values = type(self)(new_values, freq=self.freq)
+            else:
+                # fill with value
+                new_values = self.copy()
+                new_values[mask] = value
+        else:
+            new_values = self.copy()
+        return new_values
+
+    def copy(self, deep=False):
+        return type(self)(self._data.copy(), freq=self.freq)
+
+    def value_counts(self, dropna=False):
+        from pandas import Series, PeriodIndex
+
+        if dropna:
+            values = self[~self.isna()]._data
```

Comment:
.value_counts has a dropna parameter but guess can't use this as we need to convert to a primitive type and there is a separate isna check done in value_counts, hmm. ought to think about letting a ``mask`` pass thru (which we did recently with nanops.

Reply:
Sidenote: since we are considering to remove `value_counts` from the interface (https://github.com/pandas-dev/pandas/issues/22843), maybe we can already remove it here (or at least not put a lot of effort in it)